Feature: Units

benloh opened the pull request:
Addresses #60 and #51 #52 #104

## Overview

To make it easier to define and manage projects for pilot studies, we've introduced the notion of a "Unit".  A Unit defines:
* Label -- e.g. "Wetlands Unit"
* Resources -- Sources for the Evidence Library, including files like PDFs
* Ratings -- e.g. "agree", "disagree"
* Comment Types -- e.g. "Comment", "Need Evidence"

Units are defined purely via files and folders.  This should make things very portable and easy to clone, backup, and manage.
* Units are defined in relation to a classroom.  
   * Each Classroom has a "active" unit that is selected ONLY when a student creates a new model.
   * Once a model has been created, changing the classroom's "active" unit will not affect existing models
   * The general expected worfklow is: a) teacher sets up a classroom and selects a unit, b) students create models using that unit, when the unit is over, c) teacher selects a different unit for the classroom, and d) when students create a new model, it will be created with the newly selected unit
* When creating a new model, the current unit is displayed in the "Create New Model" button
* The current unit name is displayed in the lower left of the Panel Tools.
<img width="336" height="102" alt="screenshot_366" src="https://github.com/user-attachments/assets/816adde4-a3c5-475c-8537-bab271055630" />

## Changes

#### Added
- Units are now defined by folders and files in the `<repodir>/units/` folder.  See [Units Folder Structure](#units-folder-structure) below for details.
- Each classroom has a "active" unit defined, defaulting to the first unit when a classroom is created.
- Click "Reload Units" to reload ALL units to force unit definitions to reload after units have been changed.

#### Changed
- Resources are no longer served from the `/resources` folder, but instead are served directly from the `/units/<unitid>/resources` folder.

#### Removed
The following items have been removed from the admin interface:
- Ratings Definitions
- Criteria for Good Model
- Sentence Starters
- Resources



## Units Folder Structure

Units are defined as a folder. 
- You can have as many units as you want.
- Resources are stored in `resources` folders in each unit folder
- MEME will happily gulp down every `*.yaml` file and combine them all together.

```
/units
  /wetlands            # folder name works as id
    wetlands.yaml 
    /resources
      fish.pdf
      marsh.pdf
  /beavers
    beavers.yaml
    /resources
      beaver.pdf
      trees.pdf
      dam.pdf      
```

## Units YAML Data Structure

The basic unit yaml file consists of four objects: `label`, `resources`, `ratings`, and `commentTypes`.

`units/exampleUnit/example.yaml`
```yaml
label: 'Example Unit'
resources: [...]
ratings: [...]
commentTypes: [...] 
```

> [!TIP]
> Pro Tip
> Because the app will load all yaml files, you can split out the sections into separate files.
>
> e.g. 
> - `units/demo/baseunit.yaml` -- with main `label` and `ratings`
> - `units/demo/resources.yaml` -- with only the `resources` object
> - `units/demo/commentTypes.yaml` -- with only the `commentTypes` object
>
> And then you can easily copy the files to other units.
> Use caution though because if you re-define one of the objects, they can clobber each other.


#### `resources`
Resources are an array consisting of 5 items:
- `id` -- numeric id used for the footnote
- `label` -- human readable label for the evidence
- `notes` -- a short descriptive string displayed below the label in the library
- `type` -- determines the icon displayed, e.g. `idea`
- `url` -- A filename reference or a URL.

Valid `types`
```
const RESOURCE_TYPES = {
  simulation: ImageIcon,
  assumption: IdeaIcon,
  idea: IdeaIcon,
  report: DescriptionIcon,
  question: ContactSupportIcon,
  other: DescriptionIcon
};
```


Note on use:
- The `id` is used as the footnote reference.
- The order of items in the Evidence Library is determined by the order of the items in the unit definition.

e.g. "Fish" will appear first in the Evidence Library because it's the first item in the array.  When you create an evidence link for "Fish" it will use `2a` because the `id` is 2.
```yaml
resources:
  - id: 2
    label: 'Fish'
    notes: 'Aquatic creatures.'
    type: 'pdf'
    url: 'Resource 1_Habits of the New Fish.pdf'
  - id: 1
    label: 'Algae'
    notes: 'Food source.'
    type: 'pdf'
    url: 'Resource 5_Water Quality Report.pdf'
```

#### `ratings`
Ratings are an array of 3 items:
- `label` -- human readable label for the rating
- `rating` -- a numeric value, generally from positive to negative
- `svgdefKey` -- this points to the SVG icon

`svgdefKey` values:
- `ratingsAgreeStrongly`
- `ratingsAgree`
- `ratingsNone`
- `ratingsDisagree`
- `ratingsDisagreeStrongly`

```yaml
ratings:
  - label: 'Algae Agrees'
    rating: 1
    svgdefKey: 'ratingsAgree'
  - label: 'Algae None'
    rating: 0
    svgdefKey: 'ratingsNone'
  - label: 'Algae Disagrees'
    rating: -1
    svgdefKey: 'ratingsDisagree'
```

#### `commentTypes`
Comment Types are arrays with 3 items:
- `slug` -- a comment type id
- `label` -- human readable label for the comment type -- this is displayed in the comment type selector dropdown
- `prompts` -- an array of prompts that define one or more questions with either text answers or dropdown menu answers.

You can find other example comment types in the `DEFAULT_CommentTypes` constant in `dc-comments.js`.  Just copy the same format, switching JSON to YAML.

```yaml
commentTypes:
  - slug: 'cmt'
    label: 'Comment'
    prompts:
      - format: 'text'
        prompt: 'Comment'
        help: 'Use this for any general comment.'
        feedback: ''
  - slug: 'evidence'
    label: 'Backed up by enough good evidence'
    prompts:
      - format: 'dropdown'
        prompt: 'Is this based on evidence and facts?'
        options:
          - 'Yes'
          - 'Some'
          - 'No'
        helpIgnore: 'Select one.'
      - format: 'text'
        prompt: 'What would you change?'
        help: "Please list specific evidence or facts, and how you'd change the model."
        feedback: ''
```

## To Test

There have been package changes.

1. `npm ci`
2. Create units.  See [To Add a Unit](#to-add-a-unit) below.  I recommend creating at least two so that you can confirm that you are selecting a different unit.
3. `npm start`
4. Create teachers, classrooms, groups, and students as needed.
5. NOTE when a classroom is created, the first unit is selected by default.
6. Log in as a student. 
7. Create a new model.  The active unit is displayed in the "Create New Model" button.
~8. ...you should see a list of units that you can select from.~
~9. Select a unit.  The model will be created immediately.~
10. Review the Evidence Library: Are the resources you created there?
11. Open the Evidence Library and click on one of the resources to open it: Can you see the resource you added?
12. Create a new evidence link.  
13. Change the rating for the evidence link: Are the ratings you defined available and settable?
14. Create a new comment: Are the defined comment types working?
15. Close and quit the server
16. Re-start the server and re-open the project: Are evidence links and comments restored?

### Test Changing Units

1. Make sure you have at least two units defined
2. Start MEME
3. Go to `/admin` and...
   - Define two teachers
   - For the first teacher, define two classrooms
   - For the second teacher, leave the classrooms blank

We need to test a number of conditions
- What happens if we define a teacher with no classrooms?  There shouldn't be an error.
- What happens if we define a teacher with a classroom but don't bother to select a unit?  The classroom should default to the first unit.
- What happens if we switch units while a student is working on a model?  The student should continue to work with the original unit with no problems.  A newly created model should use the new unit.
- What happens if we switch units while a student is in the model select view?  A newly created model should use the new unit.
- What happens if you remove a unit to a model with an existing unit?  The app should fail gracefully?  (This requires removing the unit, then restarting the server).  Currently it falls back to the first unit.

### Test Deleting Unit

Try Deleting
Take one of the well defined classroom/units and
1. Remove the unit from the `/units` folder
2. Restart the server
3. Log in as a student: You should see the "Create New Model: Unit 'xxx' not found` button is disabled.
4. Go to admin, select the teacher and classroom for the student and select a valid unit
5. Go back to the student page: The 'Create New Model: xxx" button should be enabled with a valid unit displayed 

Try Renaming
Take one of the well defined classroom/units and
1. Rename the unit folder, e.g. `/units/example` becomes `/units/badexample`
2. Restart the server
3. Log in as a student: You should see the "Create New Model: Unit 'xxx' not found` button is disabled.
4. Go to admin, select the teacher and classroom for the student and select a valid unit
5. Go back to the student page: The 'Create New Model: xxx" button should be enabled with a valid unit displayed 


## To Add a Unit

> [!IMPORTANT]
> Always **test units** before adding them to a production server.  In general you should:
> 1. Update the unit defintion
> 2. Run `npm run dev` on a local dev machine
> 3. Review the terminal log -- if there are any validation errors, they will appear as red error messages.
> 4. Fix the validation errors
> 5. Run `npm run dev` until you clear the error
> 6. Only then should you upload to a production server.

Add units by simply adding new unit definitions and folders to the `/units/` folder.

1. Add a `<repodir>/units` folder if one doesn't already exist
2. Add a folder for the unit, e.g. `/units/wetlands`

> [!IMPORTANT]
> Keep unit folder names short with no spaces.
> The unit folder name can contain letters, numbers, underscore _, and hyphen - 

3. Add a resources folder for the unit, e.g. `units/wetlands/resources`
4. Create a `YAML` file -- it's probably easiest to copy an existing YAML file.
   The name of the yaml file actually doesn't matter.  In fact you can add multiple YAML files and they'd all load.
5. Update Label, Resources, Ratings, and CommentTypes.

> [!IMPORTANT]
> Always review terminal messages when starting the server!  
> Watch out especially for validation errors!
> If an expected Unit is missing, most likely it's because of a validation error.

6. Restart the server to load the new units
7. If you have existing Classrooms, you'll want to review the current active unit for each classroom.



## Using Units

> [!WARNING]
> Changing or **removing** unit definitions (e.g. `resources` or `commentTypes`) for existing models can
> lead to data corruption or errors or odd behaviors.  **Avoid removing existing unit information on live models!!**

A few key points:
- Units are loaded when the MEME app is started (e.g. via `npm start` or `npm run dev`).
- Currently changes to units while the server is running are not detected, so if you **make changes**, you will need to **restart the server**.
- Units are determined by a `unitId` variable that is stored with the model.  The `unitId` is the folder name.

> [!CAUTION]
> Don't arbitrarily change the unit folder name!  This might detach the model from the unit!

### Validation
The unit yaml files will do basic validation when the app loads, checking:
- Required fields are present
- Field types (string, integer, etc)
- Valid folder names (no spaces, letters, numbers, underscore, and hyphen only)

### Changing Units

You can add, modify, or remove units with some caveats, see below.  Use the Classroom Panel's "Reload Units" button to force the server to reload the unit definitions.  (You don't have to restart the server).

> [!IMPORTANT]
> Always **test units** after making any changes and fix any validation issues before adding them to a production server.  

You can safely **add** units by:
1. Define and upload your unit definitions and resources to the server
2. Open the Admin interface and in the Classrooms panel, click "Reload Units".
This will cause the server to re-read the unit definitions from disk and 

It is relatively safe to **modify** exiting unit parameters, especially things like labels:
- `label`
- resources: `label`, `notes`, `type`, `url`
- ratings: `label`, `rating`, `svgdefKey`
- commentTypes: `label`

But be VERY careful with changing ids (or `slug`).  Changing these values can lead to unexpected behavior.
- resources: Changing the `id` can lead to issues for existing models if evidence has already been linked to a resource.
- commentTypes: Adding or removing prompts for an existing comment type should be gracefully handled.  Data is preserved and shouldn't result in an error.  Though some prompts may be mapped to different prompts.

AND, be very careful **removing** any unit definition!!  It is generally not recommended and can lead to issues for any models that already make use of a feature.  The app should mostly gracefully fall back to resources that are not found, but the app will display error messages or fail silently.  There is no migration code.
- resources: Removing a resource can lead to errors if an evidence link has already been created.
- commentType: Removing a commentType should be handled gracefully, but the comment will have a "not found"n error displayed.
- ratings: Removing a rating can lead to unexpected behavior.

Adding a resource to an existing unit can work, but if the student model is open, you may have to reload the model to add new evidence links.

## Issues

> [!WARNING]
> Existing models that predate Units might be a problem.  You're best off creating a new model and immediately selecting a unit.


---

# To Do
- [x] Update wiki with Units
   - [ ] Update [Dataset Management](https://github.com/theRAPTLab/meme-2023/wiki/Dataset-Management) -- remove resources section?
- [ ] Review/Confirm Electron support -- do we still need to support it?

benloh committed:
units: lint fix

benloh committed:
units: Add `js-yaml`

benloh committed:
units: lint fix

benloh committed:
units: Add `unitId` to model data

benloh committed:
units: Load units from server

benloh committed:
units: Load client-side unit data from server

benloh committed:
units: Select unit when creating new model.

benloh committed:
units: Load Ratings from Units (not Admin DB)

benloh committed:
units: Load Resources from Units (not Admin DB)

benloh committed:
units: Show unit label in panelbar.

benloh committed:
units: Added safeguards when loading yaml to prevent function injection

benloh committed:
units: Restore resource loading when loading a saved model

benloh committed:
units: Deprecate classroomResources

benloh committed:
units: Deprecate classroomResources in adm-objects too

benloh committed:
units: Add unitId to adm-settings

benloh committed:
units: Load commentTypes from unit

benloh committed:
units: Add Unit data validation

benloh committed:
units: Validate unit folder name

benloh committed:
units: Add units README.md

benloh committed:
units: Add avj (Another JSON Validator) for validation

benloh committed:
units: Add "example" unit.

benloh committed:
units: Default to selected Model's unit

benloh committed:
units: Valid resource types are 'simulation', 'assumption', 'idea', 'report', 'question', 'other'

benloh committed:
units: Remove 'Criteria', 'SentenceStarters', 'ResourcesList', 'RatingsView' from admin interface

benloh commented:
## Problems
- [x] Initial units load might need explicit resource copying

benloh committed:
units: Serve unit resources under `/units` rather than `/resources` to allow duplicate filenames and unit folder organization

benloh committed:
units: Fix missing unitId in EVResourceItemDialog url

benloh committed:
units: Fix example unit comment type prompts

benloh commented:
@jdanish Here's a first pass at Units.  There's a lot there, so let me know if you have any questions.

One big question for you: Is electron support still needed?  If we do need it, there might be a lot of cleanup/revision still ahead.

jdanish commented:
Hey Ben thanks!! Starting to look and this is great. Two questions / thoughts:

1) I think we want the teacher to pick the current unit for all students not the individual students who will be confused by this

2) in the old admin, resources could be edited on DO via the admin interface but now we need ftp access. Is there any way to make it so it won't require ftp? I am ok with FTP but not all of our teams are.

3) related to 2, the value of electron has been being able to load archived versions for analysis by grad students. I think that challenge remains?

More soon!

benloh commented:
> 1. ...want the teacher to pick the current unit for all students

Yeah, I get why you want this.  But this is a little complicated.  Presumably a student group might be associated with more than one Unit, right?  So there isn't a one-to-one correspondence between a student/group and a Unit.  

So I think this means a teacher needs to log into the admin and select some "Active Unit" selector that designates a Unit as the unit to use whenever students create new models?  So the teacher will have to remember to do this before the class starts?  It feels wonky, but that choice clearly needs to be guided by the teacher.  So the sequence would be:

1. Teacher Logs into admin
2. Teacher selects themselves
3. Teacher selects classroom
4. The available units are displayed, with "unselected" as the default
5. Teacher selects a unit
6. Next time at student logs into create a model, the model will use the selected unit
7. When the unit is over, the teacher needs to repeat the sequence to select a different unit

Since the unit is tied to the model, once a model is created we don't have to worry about what the currently selected unit is.  The model will continue to use whichever unit they first designated.

Does that sound right?


> 2. ...need ftp access

Actually I think you always needed FTP access.  Unless I'm overlooking something, we didn't have a facility for uploading files.  When you define a resources, you're merely providing a URL.  If you're adding a file you needed to add it to the resources folder.

Anyway, with the current units implementation, you have to FTP to add a unit regardless, since you need to define files and folders directly in the code folders on the server.

I was really trying to avoid a situation where we were opening up security holes by allowing uploads.  But if you're willing to live with that, I suppose we could consider it.  Or perhaps figure out a way to just serve asset files.  I think Sri would be very unhappy with this though...On second thought this makes me really nervous and I don't think we should do it.

Can you live with something like a S3 bucket/Spaces bucket? 
- We would limit the uploads to just resource assets (e.g. pdfs, images) and not unit definitions.
- Grad students and teachers could use Transmit or CyberDuck to do the transfer
- They would each get their own account so there is at least some security
- We would limit what they are allowed to upload (e.g. a NetLogo html is probably out of the question unless you're hosting it yourself somewhere)
- You would still be defining units by hand, requiring FTP to the code base, but at least we're now isolating that to just the media assets.

Setting up the login credentials would still be a challenge though, e.g. it'd probably involve SSH keys and whatnot.

And, this might not be all that useful if you still need to define units via FTP.

Another idea--but this is now getting complicated--would be to set up a second "meme-units" repo.  You would define the units themselves as well as the resources in the repo.  This gives you history and version control.  We would then set up a github workflow that would push the changes to the server whenever changes or PRs are made in the unit. 
- Any user who is updating units would have to have a github account
- Any user who is updating might inadvertently change a server
I like the version control and history, but there's a lot of room for error.  So this is also a little fraught...


> 3. ...electron for grad students

Ah, I didn't realize that was still a big use model.  I'll have to add support for electron.

jdanish commented:
> Does that sound right?

Yes except: highlight the current unit instead of blank as default. And maybe default to first in list. So teachers with only a single unit don't do any of that.

It seems a lot, but previously the teacher would have to log in, do all that and then select specific resources. So this is actually easier.

> Always needed FTP

Yes for uploading pdfs, but for changing their names and turning them on and off we had the UI. So I could upload pdfs and then assign grad students for the rest.

I can keep doing the PDF uploads, and my guess is that it's still a huge hassle to do template editing, so ... I'll have to stick with FTP for all the things. Given where we are at on budget, I think we need to live with that for now.

Are the screen captures that are just img uploads not a security hole?  I was hoping something like that with PDFs or GIFs but the bigger issue seems to be editing the unit definition to point to them, and assuming that is not anywhere close ... let's hold off on all of it.

> Still need electron

Yeah, sorry.

benloh commented:
> I can keep doing the PDF uploads, and my guess is that it's still a huge hassle to do template editing, so ... I'll have to stick with FTP for all the things. Given where we are at on budget, I think we need to live with that for now.
> Are the screen captures that are just img uploads not a security hole? I was hoping something like that with PDFs or GIFs but the bigger issue seems to be editing the unit definition to point to them, and assuming that is not anywhere close ... let's hold off on all of it.

Yeah, at the end of the day, since you're stuck having to upload yaml files somehow in a secure manner, being able to upload PDFs and images doesn't really help.  Any way you look at it, we'd have to add a lot of security.

Actually another slightly simpler approach if you and your grad students are comfortable with github is to set up custom unit repo.  When the server starts, it'd do a pull request to grab the latest unit definitions and resource files and store it locally.  Then the rest of the app starts up normally.  So this is a little easier in that you don't need to use FTP, nor need a fancy github workflow.  But it does require being comfortable with using github.  This would probably involve adding some kind of units repo config that you need to run before the server starts, since we probably want to support a set of units for NJ and a different set for IU.

Also with the current approach, if you change a unit we have to restart the server (though I'm  looking into whether or not we can safely restart the unit loading).

jdanish commented:
I think ftp is fine given that I'd need to teach the students either way.

benloh committed:
units: Improve legibility of validation error messages

benloh committed:
units: Fix resourceTypesList typo

benloh committed:
units: Remove Unit dialog from Model Select (moved to Admin)

benloh committed:
units: Add unitId to Classroom adm-objects definition

benloh committed:
units: Add Units selector to Classroom Admin Panel

benloh committed:
units: Show selected unit on "Create New Model" button.  Disable if unit is bad.

benloh committed:
units: Clean up debug messages

benloh committed:
units: lint fix

benloh committed:
units: Restore display of Resource ID.  Replace `resource.rsrcId` with `resource.id` for units

benloh commented:
@jdanish I've added a "Unit" selector to the Classroom admin panel, so now each Classroom will have a "active" unit.  I've revised the testing instructions.  You might find some odd interactions if you're working with an existing database.  It's probably cleanest to start fresh (e.g. delete `meme.loki`).

jdanish commented:
I started fresh and I think that makes more sense now for sure, thanks!

jdanish commented:
would it be possible to add the unit name to the two table views (in admin, and in models when you first login).

jdanish commented:
Thinking about the use model, I am thinking it'd be easier to teach students to ftp and hit a button than to use console ... would an option be a "restart" button in the admin interface which either re-loads units after they've been uploaded or just restarts the server if needed?  (Obviously re-load would be nicer).

jdanish commented:
Also, can we remove "criteria for a good model" from the help page since we no-longer have that?

jdanish commented:
I can make issues for any / all if it is easier.

jdanish commented:
FYI, I anticipate wanting shared resources but have verified this url works:     `url: '../../example/resources/demo.pdf'`
 
So, all good!

benloh commented:
> add the unit name to the two table views (in admin, and in models when you first login).

Yeah that should be straightforward.


> "restart" button in the admin interface which either re-loads units after they've been uploaded or just restarts the server if needed

Yeah I had that on my list of things to try next.  I believe I may be able to just trigger a resources reload without restarting the server, but we'll see.


> remove "criteria for a good model" from the help page since we no-longer have that?

Yep, on my list for cleanup.


> FYI, I anticipate wanting shared resources but have verified this url works: url: '../../example/resources/demo.pdf'

I suppose we can make a special "Shared" or "Common" folder


# To Do
- [x] add the unit name to the two table views (in admin, and in models when you first login).
- [x] remove "criteria for a good model" from the help page since we no-longer have that?
- [x] disable "Classrooms" until Teacher is selected.
- [x] "Reload" resources
- [x] Fix Move #140

benloh committed:
units: Remove Criteria from Help

jdanish commented:
Just noticed a minor bug - when "select teacher" is selected, the option to make a new classroom is still available. You can create one and it seems to work ...  but I think we don't want that.

benloh committed:
units: Disable Classroom / Unit selection if teacher is not selected

benloh committed:
units: Add 'likert' and 'discrete-slider' to comment type validation

benloh committed:
units: Set URTable default sort column using `data`

benloh committed:
units: Show "Units" in Model List Tables.

benloh committed:
units: Add "Reload Units" feature.  Use with caution!

benloh committed:
units: lint fix

benloh commented:
@jdanish Added "Reload Unit" button to Classrooms Panel.  Please review the PR text above -- there's updated information in the "Using Units" section.  Use this VERY CAREFULLY if you have existing models because it can lead to errors.  I've tried to test most of the variations on adding, modifying, and removing items, but this is complex enough that I'm sure I have not exhausted all the combinations.

jdanish commented:
Minor issue / note: if you have 3 units (deer, beaver, example), and set example as current, then remove it, and reload units, "Select unit" is selected instead of the top unit. The kid's interface correctly indicates there is no unit and can't make a new model (so, they'd tell the teacher). So ... it's fine in that it is graceful, but ideal would be it re-sets to the top unit (deer) when example is removed and the reload is triggered.

jdanish commented:
Move function doesn't appear to be working now: https://github.com/theRAPTLab/meme-2023/issues/140

benloh commented:
> Minor issue / note: if you have 3 units (deer, beaver, example), and set example as current, then remove it, and reload units, "Select unit" is selected instead of the top unit. The kid's interface correctly indicates there is no unit and can't make a new model (so, they'd tell the teacher). So ... it's fine in that it is graceful, but ideal would be it re-sets to the top unit (deer) when example is removed and the reload is triggered.

We would need to add migration if we wanted to do this.  Since there's no migration, the selected classroom unit is not changed until you take an action in the Admin UI.  If we wanted to actively make a selection to a new default unit, we'd have to introduce a migration step that re-reads all the classrooms and reselects the currently selected (default) unit.  

Migrating the selected unit for a classroom isn't a big deal since it only affects newly created models.  So we can do that if necessary.

But we probably don't want to be migrating student models, because then it becomes difficult to un-migrate/revert the changes and starts to get into dangerous territory.  We'd have to do careful migration and error checking.

jdanish commented:
Yeah just the selected unit drop down though it's not a huge priority since it isn't a crash and easily corrected

benloh committed:
units: Remove resource checks during MoveModel.  Previously needed to account for changed resources when moving between classrooms

benloh committed:
units: Add unit information when cloning a model

benloh committed:
units: Default to first unit if selected unit was removed.  Ensure there is always a valid unit for new models.

benloh committed:
units: Update Classroom panel Unit selector if unit has been removed

benloh commented:
Minor update: If a unit is removed, when the server starts or if "Reload Units" has been triggered, classrooms will auto-select the default unit.  NOTE that we are not migrating data nor saving the newly selected default unit.  The saved selected unit will remain whichever unit was previously selected.  This allows you to restore the previous selected value if you restore the unit and at the same time ensures a valid default unit is always available for students creating new models.

See 0fd9cf15b46644a979cffcbead5e094e78614eb6
See bb771884f97fdc9b15d03368f362d383a606c416

benloh committed:
units: Any classroom with a missing unit should auto-select the default unit.

benloh commented:
Minor update: Classrooms will now save the default unit id during server load and reload if the selected unit has been removed.  (Previously, classrooms would try to continue to use the previously saved (missing) unit and it was impossible to reselect the default unit without first selecting "None" ("Select a Unit") and then reselecting the unit.)  This essentially does a crude migration of the classroom database records. See a947da1cac4e67bca2e07847f94f3a813badfc39

jdanish commented:
@benloh this feels solid enough to go ahead and merge the PR and keep going. We won't be in classrooms with this for a while.

benloh commented:
> @benloh this feels solid enough to go ahead and merge the PR and keep going. We won't be in classrooms with this for a while.

@jdanish actually if the rest is now working, we need to turn to making sure Electron will work before we finish the PR.  Unless for some reason your pilot study or there are other timing considerations that make us want to wrap that early?

jdanish commented:
Sounds great!

benloh committed:
units: Add units support for Electron.  Required lazy loading ajv because of node dependency loading.

benloh committed:
units: Lint, add doc

benloh commented:
@jdanish I have Electron working with units now (I think).

The next question is what else do we need to do?  There's three issues:
- [ ] Since we had to remove Material UI, the Electron interface is no longer displayed.  This is a minor issue, but...
- [ ] Import and Export functions are built with Material UI.  So if you still need that functionality, I'm going to have to tear out the Material UI and rebuild the page.
- [ ] I have not re-tested app-signing.  Is that still important?

These are pretty significant tasks (probably about a week's worth) so if they're not critical I'd rather skip it.

jdanish commented:
We definitely do not need the app signing at this point. I think we can live without the UI and without import / export if we can just grab the files from the data folder?

jdanish commented:
<img width="1022" height="403" alt="Screenshot 2025-10-17 at 7 45 02 AM" src="https://github.com/user-attachments/assets/0fa9df53-c8ba-4c65-b649-57859fe91a2e" />

NOTE: Seeing this error, but the app seems to run OK so not sure it matters?

jdanish commented:
Also, the electron folder still includes resources, which is not being used, so we should remove to avoid confusion?

jdanish commented:
Comment that the resources didn't load was removed as user error - I had only partially duplicated a unit so there was no resource.

jdanish commented:
@benloh OK, print doesn't seem to work in electron yet - if you click on a model it returns to the login

benloh commented:
> console errors

Yeah I thought I had this fixed, but need to revisit.  I think it’s an async issue.

I’ll remove the resources folder.

> print

Printing is broken everywhere.  I was gonna tackle that with #132

jdanish commented:
Awesome thanks

benloh committed:
units: Deprecate "resources" folder. Resources are now under individual "/units"

benloh committed:
units: Delay Electron BrowserWindow creation to wait for preload script flush to disk

Reverts 3e4e53d1429b3f83f00b90c91c5b7b6f1586b156

benloh commented:
# TO DO 
- [x] Fix preload error
- [x] Gracefully fail if unit has been removed and student clicks on model: `adm-units.js:48 Uncaught Error: GetUnit: unknown unitId 'extra’ UNITS: adm-units:48`
- [x] Hide `Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot`

benloh committed:
units: Electron: React 18 deprecated ReactDOM.render() in favor of createRoot()

Fixes `Warning: ReactDOM.render is no longer supported in React 18`

benloh committed:
units: Electron: Hide MZIP Importers with deprecation of MUI

benloh committed:
units: Electron: lint fix

benloh committed:
units: Electron: Deprecate MUI/styled-components.

benloh committed:
units: Electron: remote module is deprecated in Electron 14

Instead, we use window.UR.serverinfo which is set by console-preload.js

benloh committed:
units: Electron: Fix preload script error

Properly fix 3e4e53d1429b3f83f00b90c91c5b7b6f1586b156 and 9e5586a37ede8b932982e4d944437b33e64411c8 and

benloh committed:
units: Electron: lint fix

benloh committed:
units: Electron: Revert lazy load ajv. 3e4e53d1429b3f83f00b90c91c5b7b6f1586b156

benloh committed:
units: Electron: Restore electron display.  MZIP Import by dragging no longer works.

benloh committed:
units: Disable Model list item if unit is not found

benloh commented:
@jdanish I think this is working now.  A few notes:
- you should no longer get an error in the Electron console when you start up
- the MUI interface is somewhat restored — ended up the other errors were causing the UI not to render.  That said, the dragging function is not entirely functional, but you can use the zip export and import, keeping in mind that this only zips loki files and does not address units.
- the `/resources` folder should no longer show up in the Electron dist folders.  However, to completely remove it, you may need to remove the `/templates/_blank/resources` folder if it’s there.  If that folder is still in the blank template, the folder will get recreated with the packaging.

Restoring printing is next.

jdanish commented:
@benloh  Were there new commits too? I don't see new code. When the loki units are loaded, do they use whatever unit is present?  So we can just back that up by hand?

benloh commented:
@jdanish Sorry, forgot to push the commits.

The loki file should include all the model data, so the unit should be stored with the loki.  There’s just a potential disconnect if you don’t remember which units were associated with the loki file.

benloh commented:
Oh, one more note:
- If you remove a unit from an existing model, the model list item will be disabled.  So you’re not allowed to open a model with a bad unit.  To restore, just make sure you add the missing unit back in and Reload Units.  This seemed better than trying to open a model with bad unit and inadvertently changing bad links.

benloh commented:
Tested.  Merging.
